crates/red_knot_python_semantic/resources/mdtest/boundness_declaredness.md

```diff
@@ -0,0 +1,202 @@
+# Boundness and declaredness
+
+This document demonstrates how type-inference and diagnostics works for public symbols. We test the
+whole matrix of possible boundness and declaredness states. The current behavior is as follows (as
+demonstrated by the tests below):
+
+| **Public type**  | declared     | possibly-undeclared        | undeclared   |
+| ---------------- | ------------ | -------------------------- | ------------ |
+| bound            | `T_declared` | `T_declared \| T_inferred` | `T_inferred` |
+| possibly-unbound | `T_declared` | `T_declared \| T_inferred` | `T_inferred` |
+| unbound          | `T_declared` | `T_declared`               | `Unknown`    |
+
+| **Diagnostic**   | declared | possibly-undeclared       | undeclared          |
+| ---------------- | -------- | ------------------------- | ------------------- |
+| bound            |          |                           |                     |
+| possibly-unbound |          | `possibly-unbound-import` |                     |
+| unbound          |          |                           | `unresolved-import` |
```

Comment:
While there may be other behaviors here that we might reconsider, I think the fact that `possibly-unbound-import` diagnostic occurs in the "possibly-undeclared and possibly-unbound" case but not in the "undeclared and possibly-unbound" or "unbound and possibly-undeclared" cases is clearly a bug, I don't see any justification for it. I'm tempted to call that out more clearly with a TODO here (and/or below where those specific cases are tested.) I don't feel strongly about it, though, the existing note below is sufficient and refers to an issue that clearly discusses the problem.

Reply:
> but not in the "undeclared and possibly-unbound" or "unbound and possibly-undeclared"

I agree. In this table, when moving to the right or to the bottom, things only get worse. So if we raise a diagnostic in a particular cell, then everything in the lower-right quadrant of that cell (inclusive) should also raise a diagnostic. I added additional text, TODO-notes in the detailed examples, and question-marks in the corresponding cells in the table.